pkg/pairtree/pairtree.go

```diff
@@ -0,0 +1,243 @@
+)
+
+// IsHidden determines if a file is hidden based on its name.
+func IsHidden(name string) bool {
+	return strings.HasPrefix(name, ".")
+}
+
+// IsDirectory determines if an object is a directory
+func IsDirectory(obj fs.DirEntry) bool {
+	return obj.IsDir()
+}
+
+// GetPrefix reads the content of the file at the pairtree prefix path and returns it as a string
+func GetPrefix(ptRoot string) (string, error) {
+	path := filepath.Join(ptRoot, prefixDir)
+
+	// Open the file
+	file, err := os.Open(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// File does not exist, return empty string and no error
+			return "", nil
+		}
+		return "", err
+	}
+	defer file.Close()
+
+	// Read the file content
+	content, err := io.ReadAll(file)
+	if err != nil {
+		return "", err
+	}
+
+	// Check if the content is empty
+	if len(content) == 0 {
+		return "", error_msgs.Err1
+	}
+
+	// Return the content as a string
+	return string(content), nil
+}
+
+// CheckPTVer checks if the pairtree_version0_1 is populated
+func CheckPTVer(ptRoot string) error {
+	path := filepath.Join(ptRoot, verDir)
+	// Open the file
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	// Get file info
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	// Check if the file is empty
+	if fileInfo.Size() == 0 {
+		return error_msgs.Err2
+	} else {
+		return nil
+	}
+}
+
+// EncodeChar takes an id and converts special characters based on the pairtree spec
+func EncodeChar(id string) string {
+	var encoded strings.Builder
+
+	for _, r := range id {
+		switch r {
+		// Convert specific visible ASCII characters to their 3-character hexadecimal encoding
+		case ' ', '"', '*', '+', ',', '<', '=', '>', '?', '\\', '^', '|':
+			encoded.WriteString(fmt.Sprintf("^%02x", r))
+		// Perform specific single-character to single-character conversions
+		case '/':
+			encoded.WriteRune('=')
+		case ':':
+			encoded.WriteRune('+')
+		case '.':
+			encoded.WriteRune(',')
+		// Default case: Check if the character is outside the range of visible ASCII
+		default:
+			if r < 0x21 || r > 0x7E {
+				// Convert the character to its 3-character hexadecimal encoding
+				encoded.WriteString(fmt.Sprintf("^%02x", r))
+			} else {
+				// Keep all other characters the same
+				encoded.WriteRune(r)
+			}
+		}
+	}
+
+	return encoded.String()
+}
+
+// CreatePP creates the full pairpath given the root, id, and prefix
+func CreatePP(id, ptRoot, prefix string) (string, error) {
+	if strings.TrimSpace(ptRoot) == "" {
+		return "", error_msgs.Err3
+	}
+
+	if strings.TrimSpace(id) == "" {
+		return "", error_msgs.Err4
+	}
+
+	if strings.HasPrefix(id, prefix) {
+		// Remove the prefix from id
+		id = strings.TrimPrefix(id, prefix)
+	} else {
+		return "", fmt.Errorf("%w, id: '%s', prefix: '%s'", error_msgs.Err5, id, prefix)
+	}
+
+	ptRoot = filepath.Join(ptRoot, rootDir)
+
+	pairPath := caltech_pairtree.Encode(id)
```

Comment:
I'm guessing `func EncodeChar(id string) string {}` is necessary because `caltech_pairtree.Encode()` includes the splitting into pairs, and there isn't a separate encoding function in that library independent from that function?

Reply:
The `caltech_pairtree.Encode()` only splits into pairs and doesn't deal with or do anything with special characters which is a part of the pairtree spec. The encode and decode functions from caltech are incredibly basic and only split things into pairs and nothing else. That's why there are multiple functions that build the pairtree path outside of caltech. Caltech isn't actually encoding anything.

Reply:
We are waiting for Caltech to get back to us on making their EncodeChar() function public.